packages/@aws-cdk/app-staging-synthesizer/README.md

```diff
@@ -0,0 +1,291 @@
+    C --- I(Stack A)
+    I --- J(File Asset)
+    F(CDK App with File/Image Asset) --- G(Staging Stack B)
+    F --- K(Stack B)
+    K --- L(File Asset <br> Image Asset)
+    G --- H(FilePublishingRole <br> ImagePublishingRole <br> StagingBucket <br> ContainerAssetsRepository <br> FileAssetsBucketEncryptionKey)
+    D --- E(FilePublishingRole <br> StagingBucket <br> FileAssetsBucketEncryptionKey)
+    M(CDK App with no Assets) --- N(Stack C)
+    N --- O(<br>)
+```
+
+This allows staging resources to be created when needed next to the CDK App. It has the following
+benefits:
+
+- Bootstrapping will be faster since the heavy resource of a KMS key is no longer involved.
+- Because roles are a global resource, every account now only needs to be bootstrapped once.
+- Users have a familiar way to customize staging resources in the CDK Application.
+
+> As this library is `experimental`, the accompanying Bootstrap Stack is not yet implemented. To use this
+> library right now, you must reuse roles that have been traditionally bootstrapped.
+
+## Synthesizer
+
+To use this library, supply the `AppStagingSynthesizer` in as the default synthesizer to the app.
+This will ensure that a Staging Stack will be created next to the CDK App to hold the staging resources.
+
+`AppStagingSynthesizer` comes with static methods covering the use-cases for the synthesizer. 
+
+### Using the Default Staging Stack per Environment
+
+The most common use case will be to use the built-in default resources. In this scenario, the
+synthesizer will create a new Staging Stack in each environment the CDK App is deployed to store
+its staging resources. To use this kind of synthesizer, use `AppStagingSynthesizer.defaultResources()`.
+
+```ts
+const app = new App({
+  defaultSynthesizer: AppStagingSynthesizer.defaultResources({
+    appId: 'my-app-id',
+  }),
+});
+```
+
+Every CDK App that uses the `DefaultStagingStack` must include an `appId`. This should
+be an identifier unique to the app and is used to differentiate staging resources associated
+with the app.
+
+### Using a Custom Staging Stack per Environment
+
+Use `AppStagingSynthesizer.customFactory()` to supply a custom staging stack on a per-environment basis.
+This has the benefit of providing a custom Staging Stack that can be created in every environment the CDK App
+is deployed to.
+
+```ts
+import { IStagingResources } from '@aws-cdk/app-staging-synthesizer';
+
+class CustomStagingStack extends Stack implements IStagingResources {
+  // ...
+}
+
+const app = new App({
+  defaultSynthesizer: AppStagingSynthesizer.customFactory({
+    factory: {
+      obtainStagingResources(stack, context) {
+        const app = App.of(stack);
+        if (!App.isApp(app)) {
+          throw new Error(`Stack ${stack.stackName} must be part of an App`);
+        }
+
+        return new CustomStagingStack(app, `CustomStagingStack-${appId}-${context.environmentString}`, { appId: 'my-app-id' }),
+      },
+    },
```

Comment:
Make an explicit class that implements the factory. This is not translatable using Rosetta.